### PR TITLE
Raise and consume events for tileSource open and load.

### PR DESCRIPTION
**Pull request by nein09**
For https://github.com/sul-dlss/iiifManifestLayouts/issues/75 - what other events might we want to raise?

OpenSeadragon provides facilities for noting when the tileSource fails to load. Do we want to fall back to the thumbnail in that case? Should we raise an error? 

Note that it's based on https://github.com/sul-dlss/iiifManifestLayouts/pull/94.

### CODE REVIEW

**Comment by iangilman** [comment]
This is a great start for discussion! 

I think the first question is what kind of event mechanism do we want? Sending custom events through the browser is definitely an option. OSD has an event mechanism built into it:

http://openseadragon.github.io/docs/OpenSeadragon.EventSource.html

...so that's another option. We could also write a custom event mechanism. 

Part of this question is internal...how do we manage the events inside the library? Part is external...how do we want to expose the events to consumers of the library? 

If we go with browser events, we may want to add some sort of extra indication that they are coming from this library...like a common prefix in the event name, for instance. Also if there are two copies of this library on the same page, we need to be able to distinguish between the two. Is there an element that represents the library that can be used as the event source? 

If we go with browser events, keep in mind that dispatchEvent was introduced to IE in version 9. I assume this library does not attempt to support IE8, but we should have @aeschylus verify for us. 

Regarding falling back to thumbnails if the tileSource fails to load, I think the idea is that we would always start with thumbnail, so if the tileSource fails, all we need to do is leave the thumbnail where it is. Of course we should also send out an event so the library consumer can do something if they want to. 

As for what additional events to support, I'll leave that to @aeschylus.

File: src/canvasObject.js [hunk 1]
```diff
@@ -52,6 +53,7 @@ CanvasObject.prototype = {
             if(previousImageObj){
               viewer.world.removeItem(previousImageObj);
             }
+            document.dispatchEvent(new CustomEvent('detail-tile-source-opened', { 'detail': self.id }));
```
Comment (aeschylus) [inline src/canvasObject.js]: I don't think it's a good idea to use the native API for this. As a library, the custom events it provides should be encapsulated. So, for instance, in the "separate renderer" branch, I've attached all the events to an event emitter for the component. This eventemiiter then has to be passed in or mixed in to the component that uses it.

Comment (nein09) [inline src/canvasObject.js]: Thanks- that branch had slipped my mind when I was working on this, so I'll take a look at it as a reference, and move forward from there.

**Comment by nein09** [comment]
Okay, I've gone with the mixin/event emitter model, in a minimalistic way. @aeschylus , if you want to have a lot of events or if you want me to make the event emitter more of an entity, we could give it its own iiifManifestLayoutEvents module, or something like that.

**Comment by iangilman** [comment]
Seems reasonable to me. 

Perhaps the next step is to expose the events outside of the library, i.e. use them from https://github.com/sul-dlss/iiifManifestLayouts/blob/master/example/index.html

**Comment by nein09** [comment]
@iangilman maybe something like that? It's just a shim that adds listeners to the dispatcher.

**Comment by iangilman** [comment]
Yup, seems good :)

**Comment by iangilman** [comment]
Fixed merge conflict.

**Comment by iangilman** [comment]
@aeschylus Even though #75 is not our primary focus at the moment, maybe it would still be good to merge this patch in (after #94)? I think it's the direction we want to go, and we can build on it with future patches, but merging this much in now will keep our changes together.

**Comment by nein09** [comment]
Since the was based on #94, which has been merged in, closing this PR and opening a new one on master.

**Pull request closed without merging**